### PR TITLE
config: refactor config/theme directories handling

### PR DESCRIPTION
**Pull request by johanmalm**
Instead of caching global config/theme directories, create lists of paths (e.g. '/home/foo/.config/labwc/rc.xml', '/etc/xdg/labwc/rc.xml', etc). This gives the flexibility to iterate forwards or backwards depending on if a config merge is wanted.

It also enables better fallback, for example if a particular theme does not exist in $HOME/.local/share/themes, it can be searched for in ~/.themes/ and so on.

This has the added benefit of using `fopen()` only and avoid the additional `stat()` in most situations.

~~Note: This only half-finished, so just pushing to share the idea. The for-loops in rcxml.c and menu.c use the new approach.~~

Related-to: #1411

@spl237 @ahesford Appreciate thoughts on alternative approach to #1411 
It's a bit bit rushed and only half-finished, but gives an idea of another approach... for discussion.
Happy to carry on working on it, but might take me a few days to get back to it.

### CODE REVIEW

**Comment by ahesford** [comment]
This is a clean approach. All of the parsing logic reads the same, except the direction of iteration gets reversed and we break after the first hit when not merging. Furthermore, it has the added benefit of not caching the file paths, which will fix behavior like (open labwc with only `/etc/xdg/labwc/rc.xml` -> create `~/.config/labwc/rc.xml` -> reload labwc) failing to recognize the newly created file.

I think it's worth carrying this to completion, although we'll have to reconcile the fact that, even when merging, `$XDG_CONFIG_HOME` should *replace*, not augment, `$HOME/.config` when it is defined; and `$XDG_CONFIG_DIRS` should replace `/etc/xdg` when defined. This could be cleanly handled by extending `struct dir` as, *e.g.*,
```c
struct dir {
	const char *prefix;
	const char *default_prefix;
	const char *path;
}

static struct dir config_dirs[] {
	{ "XDG_CONFIG_HOME", "...", "labwc" },
	{ "XDG_CONFIG_DIRS", "/etc/xdg", "labwc" },
	{ NULL, NULL },
}
```
where `...` is a placeholder for the fact that the default prefix for `XDG_CONFIG_HOME` can't be a static string and must instead expand the value of `$HOME`... not sure how I'd handle that.

**Comment by spl237** [comment]
At a quick glance, looks fine to me.

I've always been of the opinion that it is far better for original authors to make changes to code than to have them via PRs -the modifications are then written with far better knowledge of the intent and implementation of the original code, and they are then easier to maintain than random third-party additions.

Happy to have my PR closed and superseded by this one.

**Comment by johanmalm** [comment]
Ready for review. 

I have given it a reasonable amount of testing during development.

Run with `LABWC_DEBUG_DIR_CONFIG_AND_THEME=1 ./build/labwc -V` for a better of idea of what's going on.

We could consider renaming `common/dir.c` to `common/paths.c` - but suggest we leave that for a separate PR.

File: src/common/dir.c [hunk 1]
```diff
@@ -4,143 +4,189 @@
...
-			ctx->build_path_fn(ctx, NULL, d.path);
+		struct buf prefix;
+		buf_init(&prefix);
+
+		/*
+		 * Replace (rather than augment) $HOME/.config with
+		 * $XDG_CONFIG_HOME if defined, and so on for the other
+		 * XDG Base Directories.
+		 */
+		buf_add(&prefix, getenv(d.prefix) ? : d.default_prefix);
```
Comment (ahesford) [inline src/common/dir.c:100]: `?:` without a second expression seems to be a GNU extension.
```suggestion
		char *pfxenv = getenv(d.prefix);
		buf_add(&prefix, pfxenv ? pfxenv : d.default_prefix);
```

**Review by ahesford — APPROVED** [review]
I haven't run-tested this, but it looks like a great approach to the config-merging problem and I only have a few minor suggestions.

File: docs/labwc-config.5.scd [hunk 2]
```diff
@@ -18,6 +18,18 @@ searched for in the following order:
 - ${XDG_CONFIG_HOME:-$HOME/.config}/labwc
 - ${XDG_CONFIG_DIRS:-/etc/xdg}/labwc
 
+When $XDG_CONFIG_HOME is defined, it replaces (rather than augments)
+$HOME/.config. The same is the case for $XDG_CONFIG_DIRS and /etc/xdg.
+
+XDG Base Dir Spec does not specify whether or not a compositor should
+(a) define that only the file under the most important base directory
+should be used, or (b) define rules for merging the information from the
+different files.
```
[the remark this answers is not in the document]
Comment (johanmalm) [reply docs/labwc-config.5.scd]: Thanks. Done.

File: docs/labwc-config.5.scd [hunk 3]
```diff
@@ -18,6 +18,18 @@ searched for in the following order:
...
+
+By default, labwc uses option (a), so looks through directories in descending
+order of importantce (most important first) and only reads the first encountered
+config file.  With the --merge-config option, this behaviour is reversed.
```
[the remark this answers is not in the document]
Comment (johanmalm) [reply docs/labwc-config.5.scd]: Done.

File: src/common/dir.c (hunk 1 above)
Comment (johanmalm) [reply src/common/dir.c]: Done.
Yes, should avoid that really.... have just become rather fond of it.

**Comment by johanmalm** [comment]
Thanks for review. Have rebased the fixups. Will go for the merge.

**Comment by spl237** [comment]
Just tested this here - all looks good to me; have tested mergeing of rc.xml, autostart and environment, and all do what I would expect.

Many thanks!

**Pull request merged**